doc/source/example-lm.rst

```diff
@@ -0,0 +1,298 @@
+                              '* n_GPUs) to be a fixed input value for different '
+                              'number of GPUs or CPUs')
+      # fmt: on
+      return group
+
+
+  def ray_main():
+      """Entrance function to the fairseq library, providing fault-tolerance."""
+
+      # Parse the command line arguments.
+      parser = options.get_training_parser()
+      add_ray_args(parser)
+      args = options.parse_args_and_arch(parser)
+      original_args = copy.deepcopy(args)
+
+      # Main loop for fault-tolerant training.
+      retry = True
+      while retry:
+          args = copy.deepcopy(original_args)
+
+          # Initialize Ray.
+          ray.init(address=args.ray_address)
+
+          # Get the number of resources and set the corresponding fields.
+          if args.cpu:
+              args.distributed_world_size = int(ray.cluster_resources()["CPU"])
+          else:
+              n_gpus = int(ray.cluster_resources().get("GPU", 0))
+              while n_gpus == 0:
+                  print("No GPUs available, wait 10 seconds")
+                  time.sleep(10)
+                  n_gpus = int(ray.cluster_resources().get("GPU", 0))
+              args.distributed_world_size = n_gpus
+
+          # Set the total batch_size to a fixed number no matter how many GPUs we
+          # will use.
+          if args.fix_batch_size is not None:
+              args.update_freq = math.ceil(
+                  args.fix_batch_size / (args.max_sentences *
+                                        args.distributed_world_size))
+              print("Training on %d GPUs, max_sentences=%d, update_freq=%d"
+                    % (args.distributed_world_size, args.max_sentences,
+                      args.fix_batch_size))
+
+          # Set up Ray distributed actors.
+          Actor = ray.remote(
+              num_cpus=1, num_gpus=int(not args.cpu))(RayDistributedActor)
+          workers = [Actor.remote() for i in range(args.distributed_world_size)]
+
+          # Get the IP address and a free port of actor 0, which is used for
+          # fairseq distributed training.
+          ip = ray.get(workers[0].get_node_ip.remote())
+          port = ray.get(workers[0].find_free_port.remote())
+          address = "tcp://{ip}:{port}".format(ip=ip, port=port)
+
+          # Start the remote processes, and check whether their are any process
+          # fails. If so, restart all the processes.
+          unfinished = [worker.run.remote(address, i, args)
+                        for i, worker in enumerate(workers)]
+          try:
+              while len(unfinished) > 0:
+                  finished, unfinished = ray.wait(unfinished)
+                  finished = ray.get(finished)
+              retry = False
+          except Exception as inst:
+              print("Ray restart because following error occurs:")
+              print(inst)
+              retry = True
+          ray.shutdown()
```

Comment:
Can you make this into a separate code block with a caption on top?

Reply:
resolved in the latest comment